### PR TITLE
Include generic interfaces in implements clause completion

### PR DESCRIPTION
**Pull request by rchande**
**Customer scenario**

Customer is trying to type an implements clause to implement a member of a generic interface. The generic interface is not listed in the completion list.

**Bugs this fixes:**

https://github.com/dotnet/roslyn/issues/18006

**Workarounds, if any**

None

**Risk**

Low, this is only a change in how we show completion within implements clauses in VB.

**Performance impact**

Low, for each symbol we might suggest we now check to see if its original definition was returned by lookup symbols (instead of the symbol itself).

**Is this a regression from a previous update?**

No

**Root cause analysis:**
Test hole, regression test has been added.

**How was the bug found?**

Customer reported.

### CODE REVIEW

**Comment by rchande** [comment]
Tag @dotnet/roslyn-ide for review

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/ImplementsClauseCompletionProvider.vb [hunk 1]
```diff
@@ -171,8 +171,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
             Next
 
             Dim symbols = semanticModel.LookupSymbols(position)
+            Dim availableInterfacesAndContainers = interfacesAndContainers.Where(Function(s) symbols.Contains(s.OriginalDefinition)).ToImmutableArray()
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/ImplementsClauseCompletionProvider.vb]: Are symbols themselves guaranteed to be an OriginalDefinition? What happens if I'm something like B(Of T) Implements IList(Of T)? or perhaps IList(Of Integer) (as a concrete unrelated type)?

Comment (sharwell) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/ImplementsClauseCompletionProvider.vb]: ❗️ 

The following operations are both O(n)

* `interfacesAndContainers.Where(...).ToImmutableArray()`
* `symbols.Contains(s.OriginalDefinition)`

These operations are nested, making the overall operation O(n²).

However, the original expression used a sequence (as opposed to nesting) of O(n) operations. This is a regression in the algorithmic complexity of the completion provider.

**Comment by rchande** [comment]
Thanks @jasonmalinowski -- added additional tests for the cases you mentioned.
@sharwell -- put symbols into a hashset for O(1) lookup.

**Comment by rchande** [comment]
Tag @MattGertz for ask mode approval

**Comment by MattGertz** [comment]
Approved pending tests and @sharwell signoff.  (I.e., scenario meets the bar.)

**Comment by rchande** [comment]
retest windows_release_vs-integration_prtest

**Pull request merged**